### PR TITLE
Add Flash Attention 2 to M2M100 model

### PR DESCRIPTION
**Pull request by visheratin**
# What does this PR do?

This PR adds support for Flash Attention 2 in M2M100 models (e.g., NLLB). [Here](https://colab.research.google.com/drive/1QaNgDMuUYAa6T_F5IrsfQzAwBA9ZwFyQ?usp=sharing) is the Colab notebook with a working demo.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker @younesbelkada

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Hi @visheratin 
Thanks for this great addition ! I see in the PR you used some old / deprecated variables such as `_use_flash_attention_2`, please see: https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L405 - you need to inherit from `M2M100Attention`. There is also sligtly more work to be done on the documentation side to add expected speedups, check out this most recent PR: https://github.com/huggingface/transformers/pull/29226 to see what are the required changes and let me know if you have any question - thanks !

**Comment by younesbelkada** [comment]
@visheratin correct, for llama it's because the attention mask logic has been refactored in favor of `AttentionMaskConverter`. OK to use `_use_flash_attention_2`, the other points are still valid though, can you check what has been done for GPT2 and make the changes accordingly? 🙏

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 1]
```diff
@@ -317,6 +324,207 @@ def forward(
...
+
+        # Use the `embed_dim` from the config (stored in the class) rather than `hidden_state` because `attn_output` can be
+        # partitioned across GPUs when using tensor-parallelism.
+        attn_output = attn_output.reshape(bsz, q_len, self.embed_dim)
+
+        attn_output = self.out_proj(attn_output)
+
+        return attn_output, None, past_key_value
+
+    def _flash_attention_forward(
```
Comment (younesbelkada) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:429]:
```suggestion
    # Copied from transformers.models.llama.modeling_llama.LlamaFlashAttention2._flash_attention_forward
    def _flash_attention_forward(
```
Can't this use copied from as well ?

**Review by younesbelkada — COMMENTED** [review]
Looks very clean ! Thanks for working on this ! I left one single comment - what do you think?

**Review by younesbelkada — APPROVED** [review]
Thanks again for the smooth integration!

File: docs/source/en/model_doc/nllb.md [hunk 2]
```diff
@@ -145,3 +145,46 @@ UN-Chef sagt, es gibt keine militärische Lösung in Syrien
...
+>>> tokenizer = AutoTokenizer.from_pretrained("facebook/nllb-200-distilled-600M")
+
+>>> article = "Şeful ONU spune că nu există o soluţie militară în Siria"
+>>> inputs = tokenizer(article, return_tensors="pt").to("cuda")
+
+>>> translated_tokens = model.generate(
+...     **inputs, forced_bos_token_id=tokenizer.lang_code_to_id["deu_Latn"], max_length=30
+... )
+>>> tokenizer.batch_decode(translated_tokens, skip_special_tokens=True)[0]
+UN-Chef sagt, es gibt keine militärische Lösung in Syrien
```
Comment (amyeroberts) [inline docs/source/en/model_doc/nllb.md:181]:
```suggestion
"UN-Chef sagt, es gibt keine militärische Lösung in Syrien"
```

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 3]
```diff
@@ -967,18 +1185,24 @@ def forward(
         if inputs_embeds is None:
             inputs_embeds = self.embed_tokens(input_ids) * self.embed_scale
 
-        # create causal mask
-        # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
-        combined_attention_mask = _prepare_4d_causal_attention_mask(
```
Comment (amyeroberts) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:972]: Why rename here from `combined_attention_mask` to `attention_mask`?

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 4]
```diff
@@ -1028,7 +1252,8 @@ def forward(
                     layer_outputs = self._gradient_checkpointing_func(
                         decoder_layer.__call__,
                         hidden_states,
-                        combined_attention_mask,
+                        # combined_attention_mask,
```
Comment (amyeroberts) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:1255]: I would be better to keep the old name though

```suggestion
```

File: src/transformers/models/m2m_100/modeling_m2m_100.py [hunk 5]
```diff
@@ -1040,7 +1265,8 @@ def forward(
                 else:
                     layer_outputs = decoder_layer(
                         hidden_states,
-                        attention_mask=combined_attention_mask,
+                        # attention_mask=combined_attention_mask,
```
Comment (amyeroberts) [inline src/transformers/models/m2m_100/modeling_m2m_100.py:1268]: Same here 
```suggestion
```

File: tests/models/m2m_100/test_modeling_m2m_100.py [hunk 6]
```diff
@@ -412,3 +416,48 @@ def test_seq_to_seq_generation(self):
...
+            num_beams=5,
+            forced_bos_token_id=tokenizer.get_lang_id("en"),
+        )
+
+        expected_en = [
+            "The NSA case highlights the total absence of intelligence debate",
+            "I think there are two levels of response from the French government.",
+            "When François Hollande calls Barack Obama or when Foreign Minister Laurent Fabius calls the U.S."
+            " Ambassador, they respond to a real discovery, which is that of the scale of U.S. surveillance on all"
+            " communications in France.",
```
Comment (amyeroberts) [inline tests/models/m2m_100/test_modeling_m2m_100.py:457]: 🕵️

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

Just some small nits to resolve before merge

File: src/transformers/models/m2m_100/modeling_m2m_100.py (hunk 5 above)
Comment (younesbelkada) [reply src/transformers/models/m2m_100/modeling_m2m_100.py:1268]: nice catch !

File: tests/models/m2m_100/test_modeling_m2m_100.py (hunk 6 above)
Comment (amyeroberts) [reply tests/models/m2m_100/test_modeling_m2m_100.py:457]: Ah, sorry, I didn't mean to be confusing, It's just that it was talking about surveillance so I thought I'd drop a wee spy

**Review by amyeroberts — APPROVED** [review]
Thanks again for adding and iterating!
